misk-hibernate/src/main/kotlin/misk/hibernate/Query.kt

```diff
@@ -1,17 +1,31 @@
 package misk.hibernate
 
+import javax.persistence.criteria.CriteriaBuilder
+import javax.persistence.criteria.Root
+import javax.persistence.criteria.Selection
 import kotlin.reflect.KClass
 
 /** Base class for SQL queries. */
 interface Query<T> {
+  /** The first result to return. Defaults to 0. */
+  var firstResult: Int
+
   /** How many rows to return. Must be -1 or in range 1..10_000. */
   var maxRows: Int
 
+  /** Constrain a query using a path known only at runtime. */
+  fun addConstraint(path: String, operator: Operator, value: Any? = null)
+  fun addOrder(path: String, asc: Boolean)
+  fun list(session: Session, projectedPaths: List<String>): List<List<Any?>>
```

Comment:
warrants more docs on the return type I think (what the nested list represents)

Reply:
Yeah good point. I added some more documentation.